### DISCUSSION TITLE
Change the discriminator system to be built into the username one

### DISCUSSION THREAD

**Discussion by subterfugue** [stoatchat/discussions#555] [Feature Suggestions]
I know this has been in discussion for a while, and has been accepted as an RFC and implemented as well, but I don't believe the original discussion was productive whatsoever, and had a lot of points that could be solved by simply having additional functionality over the old username system.

# What's the problem with discriminators?

First and foremost, discriminator systems like discord's are limited and are kind of flakey. One thing the developers have been deciding between is whether to use 4 characters for the discriminator, or 5. Also whether to include alphanumeric characters, or only numeric characters. 

The developers decided to stick with 4, but they said they "will keep the door open". They also decided to only opt for numeric characters.
Using the old traditional username system that Revolt had, such a thing is extremely easy to do for any user, by their own choice. Someone could go with only adding one number, and if it's taken, another number, etc. (If you consider this to be inconvenient, please continue reading for my proposed modification to the RFC)
You can see that any sort of discriminator system one would want to implement is already just a more limited subset of what can be done with normal usernames, and the discriminator systems just force something entirely unnecessary on people's tags. Maybe their username doesn't require 4 numbers, and thus they only need to transfer the username they picked, rather than a username they picked, and a discriminator that's forced on them-- which not only is harder to remember, but also less convenient to tell other people in the real world when you want to exchange information. 

Secondly, though almost an extension of the first point, any sort of discrimination system one wants can be replicated in a system where every username must be unique. A user can pick their own way to identify themselves from any other people who picked the username they previously wanted.
One thing that could be annoying when using a system where every username must be unique, is that people would need to manually add a string or numbers if their name is super common. But something like Revolt can ask for a "display name" the user can choose, and suggest a user name that is basically the display name, appended with however many numbers to make it unique. This system 
1. allows more freedom than the old discriminator style
2. makes it easier for users to get up and going with their preferred display name
3. alllows them to change their username perhaps to something that's less common. "John" could change his name to "JohnMunchesCereal" rather than "John#9512", which would be easier to communicate
4. is more future-proof and easier to implement than the RFC for discriminators that has been implemented.

And this is the modification I suggest to the RFC. It's a big one, and it's an entirely different system. But people have said it's already accepted and that I need to suggest improvements to the RFC. Well, this is what I've got.

Thirdly, the only service that's *ever* used a discriminator style of identification was discord. And even they switched off it-- remember how much money they used to make off people who simply wanted to get a cool tag! And they ignored that amount of money just to use that system. I think that speaks volumes. 

And there are *tons* of services right now that use the exact system I'm suggesting. Mastodon for an example, if your username is taken within an instance. There's also GitHub, Matrix, and so much more.

I believe that those services opted against this system for a reason, and that this should be something we learn from them.

To conclude, I think it depends on what people prefer. If you prefer an approach that gives every person the freedom to add their own little flavour to their usernames to make them unique, and specialize it for themselves, then this system seems like a good choice.

If you prefer an approach that forces a unified system that even limits special tags like 1234, to every username-- no matter how unique, and making it objectively more difficult to add users, then discriminators are the better choice.

**Comment by insertish** [comment]
I'm currently making lunch so I can't read / respond this whole post right now, but I wanted to mention this:

> -- remember how much money they used to make off people who simply wanted to get a cool tag! And they ignored that amount of money just to use that system. I think that speaks volumes.

I don't think this really speaks volumes when that same company prioritised profits over loyal users when rolling out this change.

**Reply by subterfugue** [reply]
Yeah, it was probably a bad example. I know that Discord is a very greedy company and rarely make the right decision, but to me it looks like they had no reason to switch to the new unique username system other than to make the platform more accessible and easier to use for people. It probably doesn't speak anything, but the point that one of the only companies that used this system migrated off of it still stands.

**Comment by subterfugue** [comment]
Oh, I forgot one more point that this suggestion doesn't cover: The RFC wants a solution to accounts that have a rare "short" name, to prevent selling/stealing accounts, which I have to say a really weird point that I don't think should be addressed, but if needed, you can limit usernames to only have 4, or 5 characters *at least*. thus, there's way way way less incentive to steal accounts, as there are way more possibilities in 4 or 5 characters. I think this is a better way to de-incentivize stealing accounts while still making it easier to exchange information and more convenient to remember usernames than discriminators.

Now, let's sort of create a table of how this solution works in comparison to the one presented in the RFC:

| Solution                                                                | Description                                                                                                           | Users have desired username | Selling disincentivized | Privacy | Usability | Difficulty |
| ----------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------- | :-------------------------: | :---------------------: | :-----: | :-------: | :--------: |
| **Discriminators**                                                      | Proposed in the RFC                                                                                                   |             ⚠️              |           ✅            |   ✅    |    ⚠️     |    Low     |
| Discriminators in Unique usernames                                                        | Proposed in this discussion                                                                                           |             ✅              |           ✅            |   ✅    |    ✅     |    Lower   |


Most of what the above columns mean is similar to the one in the RFC, except the "Users have desired username" option also has the added predicate that people can share precisely only the username they set (because as I said, it's easier to remember and more convenient to share)

And the Selling Disincentivized part I've already posed what I think is a much better solution than discriminators, but if you go with my original suggestion which didn't provide a solution, then that would also be a warning emoji.

I personally think that the system I'm proposing has realistically all the upsides of other systems, and none of the downsides of the other ones, or at least downsides that aren't as bad.

One could argue that this system is entirely different, but I would like to direct your attention to how similar @name#1234 is to @name1234 in a unique username system. In-fact, it's pretty much exactly equivalent. This suggestion is a sort of normalization of discriminations and moves them to the left when necessary, and gives the user the choice to pick a unique enough username without needing an extra useless 4 numbers.

**Comment by sharkaccino** [comment]
it truly amazes me the lengths people will go to complain about a few numbers

**Reply by subterfugue** [reply]
I believe this is what they call a hyperbole fallacy. Nevertheless, however many numbers it is, it's still a major overhaul over the pre-existing system, one that could likely be improved by this suggestion. That said, it amazes me the lengths *you* would go to to try to deem people's criticism and suggestions as invalid because this addition is "a few numbers". Every idea deems a fair discussion, and I believe this one does especially. That's other than the fact that I believe the original discussion was quite unfairly steered, as in my opinion most people seemed to prefer discriminators because Discord has them, and now it doesn't. I would prefer actual constructive criticism than hyperbole that doesn't target any part of the paragraphs I've written. Please be better.

**Reply by sharkaccino** [reply]
okay, then.

please explain to me how account stealing and "OG names" can be mitigated without discriminators. your proposed solution just increases the minimum character count, which honestly sounds counter-intuitive to the idea of freedom of choice. what if someone simply wants the username "amy"? ignoring the fact that it's an incredibly common name and would likely already taken in a system with 100% unique usernames, this person would be blocked from using it simply because it doesn't meet an arbitrary minimum character count.

another issue with traditional usernames is that in the event of a conflict, it's entirely up to the user to sort out a different username for themselves. this does technically leave you with the freedom of choice in the matter, but it's not remotely as convenient as usernames with discriminators. for example, let's look at how email addresses have panned out:
1. you go to set up a new email address.
2. you decide what the name will be.
3. oops, that one's taken. try something else.
4. you try to come up with a different username
5. oops, that one's taken. try something else.
6. repeat this 100x...

issues like this only become increasingly worse over time, as more accounts are created. discriminators are far better suited for mitigating this issue by assigning random numbers to your username automatically, which drastically reduces the effort required on the user's part. this allows people to focus more on their actual identity on the platform as opposed to trying to come up with something that nobody else has taken yet.

finally, i'm also very curious to know how you think unique usernames have similar privacy to usernames with discriminators. if you're a well-known celebrity or something who uses the same username across multiple platforms, would discriminators not make it much harder for people to find and pester you?

**Reply by subterfugue** [reply]
I apologize about the earlier message. The fallacy you've used isn't just hyperbole, it is also an appeal to ridicule. 

> please explain to me how account stealing and "OG names" can be mitigated without discriminators. your proposed solution just increases the minimum character count, which honestly sounds counter-intuitive to the idea of freedom of choice. what if someone simply wants the username "amy"? ignoring the fact that it's an incredibly common name and would likely already taken in a system with 100% unique usernames, this person would be blocked from using it simply because it doesn't meet an arbitrary minimum character count.

I have already explained the point of "would likely already be taken". The system I'm proposing here is a slight adjustment to a discriminator system that makes it more flexible and makes the discriminator an optional, but on by default, feature. This system is not only almost exactly equivalent to discriminators, but also has many many more possibilities. 

For account stealing, a person wanting to name themselves "amy" can still choose to do so by going for the username of "amy2234" for an example, and use the display name of "amy". Display names don't have to be unique and they can identify themselves by any display name they choose. There are many possibilities. The difference between a discriminator system and the one I'm presenting here is that the user is given a choice when available, but when it's not, they can opt for the discriminator system which is suggested by default.

> another issue with traditional usernames is that in the event of a conflict, it's entirely up to the user to sort out a different username for themselves. this does technically leave you with the freedom of choice in the matter, but it's not remotely as convenient as usernames with discriminators. for example, let's look at how email addresses have panned out:

Once again, in the system I'm proposing:
> One thing that could be annoying when using a system where every username must be unique, is that people would need to manually add a string or numbers if their name is super common. But something like Revolt can ask for a "display name" the user can choose, and suggest a user name that is basically the display name, appended with however many numbers to make it unique. This system

I propose that this be done automatically in software, as is done by many other pieces of software. This is why I call this a modification of the discriminator system, because this really keeps most of the code the same and allows more flexibility for the users and for the platform. 

> finally, i'm also very curious to know how you think unique usernames have similar privacy to usernames with discriminators. if you're a well-known celebrity or something who uses the same username across multiple platforms, would discriminators not make it much harder for people to find and pester you?

Once again, the system I'm proposing has all the abilities of a discriminator service and more. So such a popular person can for an example name their account "markiplier1954" or something, and have a display name of "markiplier". Thus, their account can be fairly hidden, while their chosen identity can be shown to people they communicate with.

**Comment by Rexogamer** [comment]
A few things:
1. It seems a bit weird to call this a "modification" of the discriminator system - assuming I'm reading this correctly, it seems that you just want to go back to the old system?
2. The last two paragraphs seem unnecessarily hostile - in addition, I question the assertion that the RFC was "unproductive" as it seemed to go by fairly smoothly and I didn't see much opposition to discriminators as an idea during it.
3. Platforms like Twitter, GitHub and the various Fediverse services are focused on, well, public connections and conversations - making it easy to find people is an important goal in this regard, whereas on a chat platform like Revolt or Discord the focus is about private conversations and communities, where you might want to make it harder for random people to find/add you.

**Reply by subterfugue** [reply]
That's kind of close, but not quite. I've always found it convenient when services suggested usernames that were unique per a chosen display name, and I believe that moving this system to be a part of the username is more flexible, and allows for any future changes without much overhaul. If you wish to change to discriminators of 5 characters, or make them alphanumeric, then you can easily do that at any point in the future without changing much of the code. This makes it more convenient to make these changes, and makes it less error prone for the codebase as well. 

For the second point I agree that it was unnecessarily hostile. I should have picked better wording, and I apologize profusely for this. 

For your third point, that's an extremely fair point! But I also point you to the fact that using this suggested system, people can change their usernames as much as they want, and use the universal display name to set the username they want to identify as. This is even more private than the discriminator system because if a popular person happened to leak their name#discrim, it would be more difficult to change that than it is to change in the system I'm suggesting.
